File: lib/scanner/plugins/applications/cms/wordpress.rb
# frozen_string_literal: true

module Yawast
  module Scanner
    module Plugins
      module Applications
        module CMS
          class WordPress
            # check to see if we can confirm the presence of WordPress
            def self.identify(uri)
              ret = nil

              # check for wp-login.php in the current directory
              resp = identify_by_path uri, uri.path

              if resp.nil?
                # if we don't get a hit at the current path, try under /blog/
                resp = identify_by_path uri, uri.path + 'blog/'
              end

              unless resp.nil?
                # confirmed hit
                res = resp[:result]
                ret = resp[:uri]

                # strip the file name from the path
                ret.path = ret.path.sub! 'wp-login.php', ''

                css = res[:body].scan /login.min.css\?ver=\d+\.\d+\.?\d*/

                ver = 'Unknown'
                if !css.count.zero?
                  ver = css[0].to_s.split('=')[1]
                else
                  # the current method doesn't work, fall back to an older method
                  css = res[:body].scan /load-styles.php\?[\w\,\;\=\&\%]+;ver=\d+\.\d+\.?\d*/
                  ver = css[0].to_s.split('=')[-1] unless css.count.zero?
                end

                Yawast::Utilities.puts_info "Found WordPress v#{ver} at #{ret}"
                Yawast::Shared::Output.log_value 'application', 'wordpress', 'uri', ret
                Yawast::Shared::Output.log_value 'application', 'wordpress', 'version', ver
                Yawast::Shared::Output.log_value 'application', 'wordpress', 'login_body', res[:body]
              end

              ret
            end

            def self.identify_by_path(uri, path)
              login_uri = uri.copy
              login_uri.path = path + 'wp-login.php'

              res = Yawast::Shared::Http.get_with_code login_uri

              if res[:code] == '200' && res[:body].include?('Powered by WordPress')
                return {result: res, uri: login_uri}
              else
                return nil
              end
            end

            def self.check_json_user_enum(uri)
              Yawast::Shared::Output.log_hash 'vulnerabilities',
                                              'wordpress_json_user_enum',
                                              {vulnerable: false, users: nil}

              json_uri = uri.copy
              json_uri.path = json_uri.path + 'wp-json/wp/v2/users'
              res = Yawast::Shared::Http.get_with_code json_uri

              if res[:code] == '200' && res[:body].include?('slug')
                # we have a likely hit
                users = nil
                begin
                  users = JSON.parse res[:body]
                rescue # rubocop:disable Style/RescueStandardError, Lint/HandleExceptions
                  # don't care why it failed
                end

                unless users.nil?
                  Yawast::Shared::Output.log_hash 'vulnerabilities',
                                                  'wordpress_json_user_enum',
                                                  {vulnerable: true, users: users}
                  Yawast::Utilities.puts_warn "WordPress WP-JSON User Enumeration at #{json_uri}"

                  users.each do |user|
                    Yawast::Utilities.puts_raw "ID: #{user['id']}\tUser Slug: '#{user['slug']}'\t\tUser Name: '#{user['name']}'"
                  end

                  puts
                end
              end
            end
          end
        end
      end
    end
  end
end

File: yawast/scanner/plugins/http/applications/wordpress.py
import re
from typing import Tuple, Union, List, cast
from urllib.parse import urljoin

from packaging import version
from requests import Response

from yawast.reporting.enums import Vulnerabilities
from yawast.scanner.plugins.evidence import Evidence
from yawast.scanner.plugins.http import version_checker
from yawast.scanner.plugins.result import Result
from yawast.shared import network


def identify(url: str) -> Tuple[Union[str, None], List[Result]]:
    results = []

    # find WordPress
    res, path = _identify_by_path(url, "")

    if path is None:
        res, path = _identify_by_path(url, "blog/")

    # check to see if we have a valid hit
    if path is not None:
        # we have a WordPress install, let's see if we can get a version
        body = res.text

        ver = "Unknown"
        # this works for modern versions
        m = re.search(r"login.min.css\?ver=\d+\.\d+\.?\d*", body)
        if m:
            ver = m.group(0).split("=")[1]
        else:
            # the current method doesn't work, fall back to an older method
            m = re.search(r"load-styles.php\?[\w,;=&%]+;ver=\d+\.\d+\.?\d*", body)
            if m:
                ver = m.group(0).split("=")[-1]

        # report that we found WordPress
        results.append(
            Result.from_evidence(
                Evidence.from_response(res, {"version": ver}),
                f"Found WordPress v{ver} at {path}",
                Vulnerabilities.APP_WORDPRESS_VERSION,
            )
        )

        # is this a current version?
        ver = cast(version.Version, version.parse(ver))
        curr_version = version_checker.get_latest_version("wordpress", ver)

        if curr_version is not None and curr_version > ver:
            results.append(
                Result.from_evidence(
                    Evidence.from_response(
                        res,
                        {
                            "installed_version": str(ver),
                            "current_verison": str(curr_version),
                        },
                    ),
                    f"WordPress Outdated: {ver} - Current: {curr_version}",
                    Vulnerabilities.APP_WORDPRESS_OUTDATED,
                )
            )

        return path, results
    else:
        return None, []


def check_json_user_enum(url: str) -> List[Result]:
    results = []
    target = urljoin(url, "wp-json/wp/v2/users")

    res = network.http_get(target, False)
    body = res.text

    if res.status_code < 300 and "slug" in body:
        data = res.json()

        # log the enum finding
        results.append(
            Result.from_evidence(
                Evidence.from_response(res),
                f"WordPress WP-JSON User Enumeration at {target}",
                Vulnerabilities.APP_WORDPRESS_USER_ENUM_API,
            )
        )

        # log the individual users
        for user in data:
            results.append(
                Result.from_evidence(
                    Evidence.from_response(
                        res,
                        {
                            "user_id": user["id"],
                            "user_slug": user["slug"],
                            "user_name": user["name"],
                        },
                    ),
                    f"ID: {user['id']}\tUser Slug: '{user['slug']}'\t\tUser Name: '{user['name']}'",
                    Vulnerabilities.APP_WORDPRESS_USER_FOUND,
                )
            )

    return results


def _identify_by_path(url: str, path: str) -> Tuple[Response, Union[str, None]]:
    target = urljoin(url, f"{path}wp-login.php")

    res = network.http_get(target, False)
    body = res.text

    if res.status_code == 200 and "Powered by WordPress" in body:
        return res, urljoin(url, path)
    else:
        return res, None
